### PR TITLE
Add hashing password

### PR DESCRIPTION
**Pull request by MarkiyanSHCH**
In back-end create service for hashing like in FS, add to authController new HttpPost request (Registration) also add new SP (spUsers_AddUser).
In front-end create modal for registration and Registration model.

### CODE REVIEW

File: WebAPI/Core/Services/AuthServices.cs [hunk 1]
```diff
@@ -15,14 +15,35 @@ public class AuthServices
     {
         private readonly IOptions<AuthOptions> _authOptions;
         private readonly IAuthRepository _authRepository;
+        private readonly HashingService _hashingService;
 
-        public AuthServices(IOptions<AuthOptions> authOptions, IAuthRepository authRepository)
-            => (this._authOptions, this._authRepository) = (authOptions, authRepository);
+        public AuthServices(IOptions<AuthOptions> authOptions, IAuthRepository authRepository, HashingService hashingService)
+            => (this._authOptions, this._authRepository, this._hashingService) = (authOptions, authRepository, hashingService);
```
Comment (bdemediuk) [inline WebAPI/Core/Services/AuthServices.cs:21]: Use usual indentation instep of an arrow function when you have more than 2 parameters.

**Review by bdemediuk — CHANGES_REQUESTED** [review]

File: WebAPI/Core/Services/Hashing/HashingService.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                    MinSaltSizeInBytes = 16,
+                    MinHashSizeInBytes = 32,
+                    MinIterationsCount = 10000
+                }
+            };
+
+        public HashingService(IHashingSettings settings)
+        {
+            this._settings = settings;
+            this._hashFormatter = new HashFormatter();
```
Comment (bdemediuk) [inline WebAPI/Core/Services/Hashing/HashingService.cs:31]: this._hashFormatter = new HashFormatter();

File: WebAPI/Core/Services/Hashing/HashingService.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        }
+
+        public string Hash(string input)
+        {
+            try
+            {
+                ParametersPolicy policy = this._supportedPolicies.GetValueOrDefault(this._settings.HashAlgorithmName);
```
Comment (bdemediuk) [inline WebAPI/Core/Services/Hashing/HashingService.cs:38]: Why do you need this?

File: WebAPI/Core/Services/Hashing/HashingService.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                if (policy == null) return null;
```
Comment (bdemediuk) [inline WebAPI/Core/Services/Hashing/HashingService.cs:39]: Use the seme formate for this

File: WebAPI/Core/Services/Hashing/HashingService.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+                using var hashAlgorithm = new Rfc2898DeriveBytes(
+                    password: input,
```
Comment (bdemediuk) [inline WebAPI/Core/Services/Hashing/HashingService.cs:42]: Why do you need pass named params?

File: WebAPI/Core/Services/Hashing/HashingService.cs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                    SaltSizeInBytes = this._settings.SaltSizeInBytes,
+                    HashSizeInBytes = this._settings.HashSizeInBytes,
+                    Iterations = this._settings.IterationsCount,
+                    Salt = salt,
+                    Key = key
+                });
+
+                return hash;
+            }
+            catch (Exception ex)
```
Comment (bdemediuk) [inline WebAPI/Core/Services/Hashing/HashingService.cs:64]: Why do you need ex?

File: WebAPI/Core/Services/Hashing/HashingService.cs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            {
+                return null;
+            }
+        }
+
+        public bool Check(string hash, string input)
```
Comment (bdemediuk) [inline WebAPI/Core/Services/Hashing/HashingService.cs:70]: Add validation

File: WebAPI/Core/Services/Hashing/HashingService.cs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+                using var hashAlgorithm = new Rfc2898DeriveBytes(
+                  password: input,
+                  salt: hashedSalt,
+                  iterations: hashContext.Iterations,
+                  hashAlgorithm: hashContext.HashAlgorithm);
+
+                byte[] inputKey = hashAlgorithm.GetBytes(hashContext.HashSizeInBytes);
+
+                bool verified = inputKey.SequenceEqual(hashedKey);
```
Comment (bdemediuk) [inline WebAPI/Core/Services/Hashing/HashingService.cs:88]: Why do you need `verified`?

File: WebAPI/Data/Programmability/Stored Procedures/spUsers_AddUser.sql [hunk 9]
```diff
@@ -0,0 +1,47 @@
+﻿USE [ProductDB]
+GO
+
+------------------------------------------------------------------------------
+IF EXISTS(
+    SELECT *
+    FROM INFORMATION_SCHEMA.ROUTINES
+    WHERE [ROUTINE_NAME] = 'spUsers_AddUser'
```
Comment (bdemediuk) [inline WebAPI/Data/Programmability/Stored Procedures/spUsers_AddUser.sql:8]: `spUsers_Add`

File: WebAPI/WebAPI/Models/PostRegistrationAccountRequest.cs [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+namespace WebAPI.Models
+{
+    public class PostRegistrationAccountRequest
+    {
+        [Required]
+        public string Name { get; set; }
+        [Required]
+        public string Email { get; set; }
+        [Required]
+        public string Password { get; set; }
```
Comment (bdemediuk) [inline WebAPI/WebAPI/Models/PostRegistrationAccountRequest.cs:12]: Add validation

File: WebAPI/WebAPI/Startup.cs [hunk 11]
```diff
@@ -5,28 +5,27 @@
...
-        }
-
-        public IConfiguration Configuration { get; }
+            => _configuration = configuration;
 
         // This method gets called by the runtime. Use this method to add services to the container.
         public void ConfigureServices(IServiceCollection services)
         {
-            var authOptions = Configuration.GetSection("Auth").Get<AuthOptions>();
+            var authOptions = _configuration.GetSection("Auth").Get<AuthOptions>();
```
Comment (bdemediuk) [inline WebAPI/WebAPI/Startup.cs:26]: Add this

File: WebAPI/WebAPI/Startup.cs [hunk 12]
```diff
@@ -5,28 +5,27 @@
...
 
-            services.Configure<AuthOptions>(Configuration.GetSection("Auth"));
+            services.Configure<AuthOptions>(_configuration.GetSection("Auth"));
```
Comment (bdemediuk) [inline WebAPI/WebAPI/Startup.cs:28]: Add this

**Review by bdemediuk — CHANGES_REQUESTED** [review]

File: WebAPI/Core/Services/Hashing/HashingService.cs [hunk 13]
```diff
@@ -71,25 +55,27 @@ public bool Check(string hash, string input)
         {
             try
             {
-                HashContext hashContext = this._hashFormatter.Parse(hash);
-                if(hashContext == null) return false;
+                if (string.IsNullOrWhiteSpace(hash) || string.IsNullOrWhiteSpace(input))
+                    throw new ArgumentNullException();
```
Comment (bdemediuk) [inline WebAPI/Core/Services/Hashing/HashingService.cs:59]: ArgumentException

File: WebAPI/WebAPI/Models/PostRegistrationAccountRequest.cs [hunk 14]
```diff
@@ -7,8 +7,10 @@ public class PostRegistrationAccountRequest
         [Required]
```
Comment (bdemediuk) [inline WebAPI/WebAPI/Models/PostRegistrationAccountRequest.cs:7]: More validation

**Review by bdemediuk — CHANGES_REQUESTED** [review]

**Review by BodyDuk — APPROVED** [review]

**Pull request merged**